Expose model bounding box to JS

achan00 opened the pull request:
#### 20780c1b3ad229ddd62aea26af9b877e1b72421e
<pre>
Expose model bounding box to JS
<a href="https://bugs.webkit.org/show_bug.cgi?id=277433">https://bugs.webkit.org/show_bug.cgi?id=277433</a>
<a href="https://rdar.apple.com/125586529">rdar://125586529</a>

Reviewed by Mike Wyrzykowski.

Add two new data members m_boundingBoxCenter and m_boundingBoxExtents
to HTMLModelElement that tracks the center and extents of the loaded model.

When the model process finishes loading the model source, it will send
the model&apos;s center and extents info in the DidFinishLoading message.
ModelProcessModelPlayer then calls ModelPlayerClient::didUpdateBoundingBox()
to pass that info along to the model element.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::m_boundingBoxExtents):
(WebCore::HTMLModelElement::setSourceURL):
(WebCore::HTMLModelElement::createModelPlayer):
(WebCore::HTMLModelElement::boundingBoxCenter const):
(WebCore::HTMLModelElement::boundingBoxExtents const):
(WebCore::HTMLModelElement::didUpdateBoundingBox):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/HTMLModelElement.idl:
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebCore/dom/DOMPointReadOnly.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::didFinishLoading):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.messages.in:

Canonical link: <a href="https://commits.webkit.org/281722@main">https://commits.webkit.org/281722@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3c41f79a7c29758b41ba23422a6165bb06060094

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60622 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39981 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13198 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64553 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11169 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62752 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47657 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11444 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49041 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7764 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62655 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37243 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52519 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29869 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33934 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9746 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10082 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55794 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10046 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66282 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4566 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9901 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56413 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4587 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52490 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56590 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13557 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3789 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35786 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36868 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37961 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36613 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -294,6 +300,8 @@ void HTMLModelElement::createModelPlayer()
     ASSERT(document().page());
 #if ENABLE(MODEL_PROCESS)
     m_entityTransform = DOMMatrixReadOnly::create(TransformationMatrix::identity, DOMMatrixReadOnly::Is2D::No);
+    m_boundingBoxCenter = DOMPointReadOnly::create(0, 0, 0, 1);
```
this is equivalent to:
```
DOMPointReadOnly::create({ })
```
as DOMPointInit is defined as:
```
struct DOMPointInit {
    double x { 0 };
    double y { 0 };
    double z { 0 };
    double w { 1 };
};
```

mwyrzykowski commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -417,6 +425,22 @@ void HTMLModelElement::didUpdateEntityTransform(ModelPlayer&, const Transformati
+
+void HTMLModelElement::didUpdateBoundingBox(ModelPlayer&, const FloatPoint3D& center, const FloatPoint3D& extents)
+{
+    m_boundingBoxCenter = DOMPointReadOnly::create(center.x(), center.y(), center.z(), 1);
```
would be nice to add a constructor to DOMPointReadOnly:
```
static Ref<DOMPointReadOnly> create(const FloatPoint3D& p) { return adoptRef(*new DOMPointReadOnly(p.x(), p.y(), p.z(), 1.)); }
```
so you can write:
```
 DOMPointReadOnly::create(center);
```

eddydas commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -383,7 +384,7 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
     startAnimating();
 
-    send(Messages::ModelProcessModelPlayer::DidFinishLoading());
+    send(Messages::ModelProcessModelPlayer::DidFinishLoading(WebCore::FloatPoint3D(m_originalBoundingBoxCenter.x, m_originalBoundingBoxCenter.y, m_originalBoundingBoxCenter.z), WebCore::FloatPoint3D(m_originalBoundingBoxExtents.x, m_originalBoundingBoxExtents.y, m_originalBoundingBoxExtents.z)));
```
Do we have a facility to create a FloatPoint3D directly from a simd_float3? Might be a good one to add

eddydas commented on Source/WebCore/Modules/model-element/HTMLModelElement.idl:
```diff
@@ -40,6 +40,9 @@
 
     [Reflect] attribute boolean interactive;
 
+    [Conditional=MODEL_PROCESS, EnabledBySetting=ModelProcessEnabled, CEReactions=NotNeeded] readonly attribute DOMPointReadOnly boundingBoxCenter;
```
I'm a little uneasy on having them as 2 separate attributes. Have you considered making it a single `boundingBox` with `{center: domPoint1, extent: domPoint2}` ?

achan00 committed:
Expose model bounding box to JS
https://bugs.webkit.org/show_bug.cgi?id=277433
rdar://125586529

Reviewed by Mike Wyrzykowski.

Add two new data members m_boundingBoxCenter and m_boundingBoxExtents
to HTMLModelElement that tracks the center and extents of the loaded model.

When the model process finishes loading the model source, it will send
the model's center and extents info in the DidFinishLoading message.
ModelProcessModelPlayer then calls ModelPlayerClient::didUpdateBoundingBox()
to pass that info along to the model element.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::m_boundingBoxExtents):
(WebCore::HTMLModelElement::setSourceURL):
(WebCore::HTMLModelElement::createModelPlayer):
(WebCore::HTMLModelElement::boundingBoxCenter const):
(WebCore::HTMLModelElement::boundingBoxExtents const):
(WebCore::HTMLModelElement::didUpdateBoundingBox):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/HTMLModelElement.idl:
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebCore/dom/DOMPointReadOnly.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::didFinishLoading):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.messages.in:

Canonical link: https://commits.webkit.org/281722@main
